server/coordinator/procedure/scatter.go

```diff
@@ -0,0 +1,199 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+package procedure
+
+import (
+	"context"
+	"sync"
+
+	"github.com/CeresDB/ceresdbproto/pkg/clusterpb"
+	"github.com/CeresDB/ceresdbproto/pkg/metaservicepb"
+	"github.com/CeresDB/ceresmeta/server/cluster"
+	"github.com/CeresDB/ceresmeta/server/coordinator/procedure/dispatch"
+	"github.com/looplab/fsm"
+	"github.com/pkg/errors"
+)
+
+const (
+	EventScatterPrepare = "EventScatterPrepare"
+	EventScatterFailed  = "EventScatterFailed"
+	EventScatterSuccess = "EventScatterSuccess"
+
+	StateScatterBegin   = "StateScatterBegin"
+	StateScatterWaiting = "StateScatterWaiting"
+	StateScatterFinish  = "StateScatterFinish"
+	StateScatterFailed  = "StateScatterFailed"
+)
+
+var (
+	scatterEvents = fsm.Events{
+		{Name: EventScatterPrepare, Src: []string{StateScatterBegin}, Dst: StateScatterWaiting},
+		{Name: EventScatterSuccess, Src: []string{StateScatterWaiting}, Dst: StateScatterFinish},
+		{Name: EventScatterFailed, Src: []string{StateScatterWaiting}, Dst: StateScatterFailed},
+	}
+	scatterCallbacks = fsm.Callbacks{
+		EventScatterPrepare: scatterPrepareCallback,
+		EventScatterFailed:  scatterFailedCallback,
+		EventScatterSuccess: scatterSuccessCallback,
+	}
+)
+
+func scatterPrepareCallback(event *fsm.Event) {
+	request := event.Args[0].(*ScatterCallbackRequest)
+	requestCluster := request.cluster
+	requestDispatch := request.dispatch
+	ctx := request.cxt
+	nodeInfo := request.nodeInfo
+
+	if requestCluster.GetClusterState() == clusterpb.ClusterTopology_STABLE {
```

Comment:
I have some doubts about the scatter procedure starts for every registered node, rather than for all the nodes, and the allocation of shards seems not determined, which is confusing.

Reply:
I have referenced scatter logic in last version and rewritten it as a Procedure. Maybe scatter procedure should not process every registered node. Before create scatter procedure, there should be a separate module to determine whether it is satisfied to create scatter procedure. This needs to be discussed again.